Measure glyphs in a single CoreText call rather than repeated calls

brentfulgham opened the pull request:
#### 9c7a8b92644023967579ef9b941cc94fba9009f6
<pre>
Measure glyphs in a single CoreText call rather than repeated calls
<a href="https://bugs.webkit.org/show_bug.cgi?id=288780">https://bugs.webkit.org/show_bug.cgi?id=288780</a>
<a href="https://rdar.apple.com/145797633">rdar://145797633</a>

Reviewed by Ryan Reno.

Rather than calling `CTFontGetBoundingRectsForGlyphs` once for each glyph (and passing a
single `CGRect` to get the size), collect all glyphs needing measurement and ask CoreText
to measure the entire set of glyphs in a single API call.

In theory, this will avoid the repeated work of creating graphics contexts, selecting a
specific font face into it, transforming, etc., before making the measurement.

* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::enclosingGlyphBoundsForTextRun): Adopt bulk font measuring.
(WebCore::ComplexTextController::adjustGlyphsAndAdvances): Ditto.
* Source/WebCore/platform/graphics/Font.h:
(WebCore::Font::boundsForGlyphs const): Added.
* Source/WebCore/platform/graphics/Glyph.h:Use #pragma once
* Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
(WebCore::DrawGlyphsRecorder::drawOTSVGRun): Adopt bulk font measuring.
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::platformBoundsForGlyphs const): Added.

Canonical link: <a href="https://commits.webkit.org/292198@main">https://commits.webkit.org/292198@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/010314fb3f2969cf53cd77566e2429d4b13c87af

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95189 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14789 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4647 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100229 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45690 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97237 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15077 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23217 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72603 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29883 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98192 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11271 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85945 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52934 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10981 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3678 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45028 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81168 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3774 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102270 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22237 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16227 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81604 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22485 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81961 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81001 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/20263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25562 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2966 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15490 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15297 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22207 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27333 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21866 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25340 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23605 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

vitorroriz commented on Source/WebCore/platform/graphics/Font.h:
```diff
@@ -425,6 +429,52 @@ ALWAYS_INLINE FloatRect Font::boundsForGlyph(Glyph glyph) const
+    if (!m_glyphToBoundsMap)
+        m_glyphToBoundsMap = makeUnique<GlyphMetricsMap<FloatRect>>();
+
+    const auto& measuredBounds = platformBoundsForGlyphs(glyphsNeedingMeasurement);
```
I believe `platformBoundsForGlyphs` is defined only for CoreText but the caller here `boundsForGlyphs` is defined for all platforms.

brentfulgham replied:
Yes -- you are right. The bots confirm I did not handle this properly. I'll revise.

vitorroriz commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -780,6 +780,22 @@ FloatRect Font::platformBoundsForGlyph(Glyph glyph) const
+Vector<FloatRect> Font::platformBoundsForGlyphs(const Vector<Glyph>& glyphs) const
+{
+    Vector<CGRect> rectsForGlyphs(glyphs.size());
+    auto ignoredBoundingBox = CTFontGetBoundingRectsForGlyphs(getCTFont(), platformData().orientation() == FontOrientation::Vertical ? kCTFontOrientationVertical : kCTFontOrientationHorizontal, glyphs.data(), rectsForGlyphs.data(), rectsForGlyphs.size());
```
I believe we are not using `ignoredBoundingBox` so probably we don't need to store it.

vitorroriz commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -780,6 +780,22 @@ FloatRect Font::platformBoundsForGlyph(Glyph glyph) const
+{
+    Vector<CGRect> rectsForGlyphs(glyphs.size());
+    auto ignoredBoundingBox = CTFontGetBoundingRectsForGlyphs(getCTFont(), platformData().orientation() == FontOrientation::Vertical ? kCTFontOrientationVertical : kCTFontOrientationHorizontal, glyphs.data(), rectsForGlyphs.data(), rectsForGlyphs.size());
+    (void)ignoredBoundingBox;
```
If we still want to store `ignoredBoundingBox` I think a better approach here could be `UNUSED_PARAM`.

vitorroriz commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -780,6 +780,22 @@ FloatRect Font::platformBoundsForGlyph(Glyph glyph) const
+    auto ignoredBoundingBox = CTFontGetBoundingRectsForGlyphs(getCTFont(), platformData().orientation() == FontOrientation::Vertical ? kCTFontOrientationVertical : kCTFontOrientationHorizontal, glyphs.data(), rectsForGlyphs.data(), rectsForGlyphs.size());
+    (void)ignoredBoundingBox;
+
+    auto boundingBoxes = rectsForGlyphs.map([&](const CGRect& cgrect) -> FloatRect {
```
nit: I think we should be able to deduce type `CGRect` here with `auto` for the lambda argument

rreno commented on Source/WebCore/platform/graphics/Font.h:
```diff
@@ -425,6 +431,56 @@ ALWAYS_INLINE FloatRect Font::boundsForGlyph(Glyph glyph) const
+    Vector<Glyph> glyphsNeedingMeasurement;
+    glyphsNeedingMeasurement.reserveInitialCapacity(glyphs.size());
+
+    Vector<size_t> positionsNeedingMeasurement;
```
What if you do something like this to cut down on bookkeeping Vectors?

```C++
struct GlyphAndPosition {
    Glyph glyph;
    size_t position;
};

Vector<GlyphAndPosition> glyphsNeedingMeasurement;
```


You could also use a std::pair. but I like named field lookups rather than `foo.first` and `foo.second`.

brentfulgham replied:
I'll give it a try!

vitorroriz commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -698,6 +709,13 @@ void ComplexTextController::adjustGlyphsAndAdvances()
+        // Measure all of the glyphs at once. We don't hold the result because
+        // the later individual lookup will retrieve it from the cache.
+        if (glyphCount > 1)
+            (void)font.boundsForGlyphs(glyphs);
```
is there a reason why we c-cast it here? I think the function is not marked as nodiscard

vitorroriz replied:
nit: Maybe it would be a good idea to call it `cache...` or `cached...` to make it clearer what the comment states.

rreno commented on Source/WebCore/platform/graphics/Font.h:
```diff
@@ -425,6 +431,62 @@ ALWAYS_INLINE FloatRect Font::boundsForGlyph(Glyph glyph) const
+
+    const auto& measuredBounds = platformBoundsForGlyphs(glyphsNeedingMeasurement);
+
+    ASSERT(measuredBounds.size() == glyphsNeedingMeasurement.size());
```
If these don't match what should we do in a release build? I think it might be best to release assert to avoid OOB access into `glyphsNeedingMeasurement` and `glyphBounds`.

rreno commented on Source/WebCore/platform/graphics/Font.h:
```diff
@@ -425,6 +431,62 @@ ALWAYS_INLINE FloatRect Font::boundsForGlyph(Glyph glyph) const
+
+    ASSERT(measuredBounds.size() == glyphsNeedingMeasurement.size());
+    size_t index = 0;
+    for (auto&& bounds : measuredBounds) {
```
is it `auto&&` so you could try to `WTFMove` into `glyphBounds`? I wonder why that didn't work. Maybe just `auto&` would be enough to let you move it?

brentfulgham committed:
Measure glyphs in a single CoreText call rather than repeated calls
https://bugs.webkit.org/show_bug.cgi?id=288780
rdar://145797633

Reviewed by Ryan Reno.

Rather than calling `CTFontGetBoundingRectsForGlyphs` once for each glyph (and passing a
single `CGRect` to get the size), collect all glyphs needing measurement and ask CoreText
to measure the entire set of glyphs in a single API call.

In theory, this will avoid the repeated work of creating graphics contexts, selecting a
specific font face into it, transforming, etc., before making the measurement.

* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::enclosingGlyphBoundsForTextRun): Adopt bulk font measuring.
(WebCore::ComplexTextController::adjustGlyphsAndAdvances): Ditto.
* Source/WebCore/platform/graphics/Font.h:
(WebCore::Font::boundsForGlyphs const): Added.
* Source/WebCore/platform/graphics/Glyph.h:Use #pragma once
* Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
(WebCore::DrawGlyphsRecorder::drawOTSVGRun): Adopt bulk font measuring.
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::platformBoundsForGlyphs const): Added.

Canonical link: https://commits.webkit.org/292198@main
